### PR TITLE
Allow device in developer mode to access team library

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## Description

NOTE: This is part 1 of 2 PRs to enable the feature described in #2294 

1. passes a boolean "licensed" flag to device in status updates
   1. device uses this to determine if the settings file should include the code to load the plugin
2. removes the `if (request.teamMembership) {` test from the pre-handlers to ensure `needsPermission` is called.
   1. Device uses project session token to authenticate (i.e. `user` is not present)
3. Update the implicit token scopes to ensure access is permitted
4. adds tests
   1. `Add to Library from Devices`
   2. `Prevents Library access from Device in different team`
   3. `Device state` -> `unlicensed` -> `gets live state`
   4. `Device state` -> `licensed` -> `gets live state`
4. updates tests
   1. various tests updated to ensure presence of `licensed` flag in state

NOTE: Docs do not mention devices _cannot access the team library_ so no docs have been updated at this point

## Related Issue(s)

#2294 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: forge/routes/auth/permissions.js [hunk 1]
```diff
@@ -7,13 +7,19 @@ const { Roles } = require('../../lib/roles.js')
...
     ],
     project: [
         'user:read',
         'project:flows:view',
         'project:flows:edit',
-        'team:projects:list'
+        'team:projects:list',
+        'library:entry:create',
+        'library:entry:list',
+        'library:entry:delete'
```
Comment (knolleary) [inline forge/routes/auth/permissions.js:22]: Were these changes necessary? We aren't changing any behaviour for the managed instances that use project-scope tokens. They access the shared library using a user-token based on who is logged in. The project-scope token should not have these permissions.

In a future iteration we will want to update the device editor to use user-scoped tokens for this.

**Review by knolleary — CHANGES_REQUESTED** [review]
I think the changes to the Project implicit scope should be reverted - unless I've missed a reason for them to have been changed.

**Comment by Steve-Mcl** [comment]
> I think the changes to the Project implicit scope should be reverted - unless I've missed a reason for them to have been changed.

They were required - but I admit now to thinking there may be a different issue to solve. 

I also replied to the suggestion [here](https://github.com/flowforge/flowforge/pull/2472#discussion_r1274758896)

I will feedback shortly.

**Comment by Steve-Mcl** [comment]
ok, so this took me a while to get back to it - my apologies.

### From a project/instance

It is the team lib plug in that makes the request to team lib. 
https://github.com/flowforge/flowforge-nr-launcher/blob/main/lib/storage/libraryPlugin.js#L21-L26 and it uses a bearer token

Before these modifications, the plugin got a free ride through the `preHander` [here ](https://github.com/flowforge/flowforge/pull/2472/files#diff-06f5b933a416234b5ab9be70426962fcd6c0e1e4c1af7e9aa2fd34d3b994f531L42-L44)and [here](https://github.com/flowforge/flowforge/pull/2472/files#diff-06f5b933a416234b5ab9be70426962fcd6c0e1e4c1af7e9aa2fd34d3b994f531L109-L111) because the `needsPermission` was wrapped in a test of `if (request.teamMembership) { ... }`

With these mods, the `needsPermission` MUST be called & therefore project types _needed_ implicit permissions (which is better than the free ride it was/is getting now IMO).

Does that make sense?

**Comment by Steve-Mcl** [comment]
@hardillb in absence of Nick, can I lean on you for a review please?

Happy to discuss the issue Nick raised and the thoughts behind his comments (they related to a future where user tokens with specific, limited access instead bearer tokens) for RBACs 

As this PR stands, to permit the device access - it was necessary to ensure `needsPermission` was called for the routes and this lead to discovering projects/instances were never actually checking permissions. This then leads to the need, in this iteration to add the implicit permissions for both the device and the project.

If any of this is unclear, I am more than happy to step through it with you.

Ta.

**Comment by hardillb** [comment]
First pass looks ok, Can we discuss it in the morning, just to make sure I've got everything?

**Review by hardillb — APPROVED** [review]
Merge once `library:entry:delete` checked

**Comment by Steve-Mcl** [comment]
> Merge once `library:entry:delete` checked

Checked & fails due to invalid token. 

While the end point is seemingly open to use (when a valid bearer is present), the plug in never generates a call to node-red for deletion (there is no code to support it) so the request is not made via NR client to NR server through to forge.

For completeness and piece of mind, I am going to delete the implicit permission regardless.

Will run a batch of manual checks to ensure that doesn't break anything.

**Comment by Steve-Mcl** [comment]
@hardillb tests on local all good with the implicit scope for deletion removed (though I did find a different UI bug!)

Could you verify/merge pls?

**Review by hardillb — APPROVED** [review]

**Pull request merged**